components/places/src/match_impl.rs

```diff
@@ -0,0 +1,483 @@
+            MatchBehavior::Anywhere | MatchBehavior::AnywhereUnmodified => find_anywhere,
+            MatchBehavior::Beginning => find_beginning,
+            MatchBehavior::BeginningCaseSensitive => find_beginning_case_sensitive,
+            _ => find_on_boundary,
+        }
+    }
+
+    fn fixup_url_str<'a>(&self, mut s: &'a str) -> Cow<'a, str> {
+        if self.match_behavior != MatchBehavior::AnywhereUnmodified {
+            if s.starts_with("http://") { s = &s[7..]; }
+            else if s.starts_with("https://") { s = &s[8..]; }
+            else if s.starts_with("ftp://") { s = &s[6..]; }
+        }
+        // TODO: would be nice to decode punycode here too, but for now
+        // this is probably fine.
+        match percent_encoding::percent_decode(s.as_bytes()).decode_utf8() {
+            Err(_) => Cow::Borrowed(s),
+            Ok(decoded) => decoded,
+        }
+    }
+
+    #[inline]
+    fn has_behavior(&self, behavior: SearchBehavior) -> bool {
+        self.search_behavior.intersects(behavior)
+    }
+
+    pub fn invoke(&self) -> bool {
+        // We only want to filter javascript: URLs if we are not supposed to search
+        // for them, and the search does not start with "javascript:".
+        if self.match_behavior == MatchBehavior::AnywhereUnmodified &&
+           self.url_str.starts_with("javascript:") &&
+           !self.has_behavior(SearchBehavior::JAVASCRIPT) &&
+           !self.search_str.starts_with("javascript:")
+        {
+            return false;
+        }
+        let matches = if self.has_behavior(SearchBehavior::RESTRICT) {
+            (!self.has_behavior(SearchBehavior::HISTORY) || self.visit_count > 0) &&
+            (!self.has_behavior(SearchBehavior::TYPED) || self.typed) &&
+            (!self.has_behavior(SearchBehavior::BOOKMARK) || self.bookmarked) &&
+            (!self.has_behavior(SearchBehavior::TAG) || !self.tags.is_empty()) &&
+            (!self.has_behavior(SearchBehavior::OPENPAGE) || self.open_page_count > 0)
+        } else {
+            (self.has_behavior(SearchBehavior::HISTORY) && self.visit_count > 0) ||
+            (self.has_behavior(SearchBehavior::TYPED) && self.typed) ||
+            (self.has_behavior(SearchBehavior::BOOKMARK) && self.bookmarked) ||
+            (self.has_behavior(SearchBehavior::TAG) && !self.tags.is_empty()) ||
+            (self.has_behavior(SearchBehavior::OPENPAGE) && self.open_page_count > 0)
+        };
+        if !matches {
+            return false;
+        }
+        let fixed_url = self.fixup_url_str(self.url_str);
+        let search_fn = self.get_search_fn();
+
+        let trimmed_url = util::slice_up_to(fixed_url.as_ref(), MAX_CHARS_TO_SEARCH_THROUGH);
+        let trimmed_title = util::slice_up_to(self.title_str, MAX_CHARS_TO_SEARCH_THROUGH);
+        for token in ascii_words(self.search_str) {
+            let matches = match (self.has_behavior(SearchBehavior::TITLE), self.has_behavior(SearchBehavior::URL)) {
```

Comment:
Woof. This is a lot easier to follow than the `else if` chain in the C++, but _gosh_ is it subtle. And we always match tags even if `SearchBehavior::TAG` isn't set, _unless_ we're matching on just the URL. 😭